src/ripper/ripcddialog.cpp

```diff
@@ -324,10 +335,33 @@ QString RipCDDialog::ParseFileFormatString(const QString& file_format,
 }
 
 void RipCDDialog::ResetDialog() {
+  QMutexLocker l(&mutex_);
   ui_->tableWidget->setRowCount(0);
   ui_->albumLineEdit->clear();
   ui_->artistLineEdit->clear();
   ui_->genreLineEdit->clear();
   ui_->yearLineEdit->clear();
   ui_->discLineEdit->clear();
 }
+
+RipCDDialog::MediaChangedPollingThread::MediaChangedPollingThread(
+    RipCDDialog& dialog)
+    : QThread(&dialog), dialog_(dialog) {}
+
+RipCDDialog::MediaChangedPollingThread::~MediaChangedPollingThread() {
+  requestInterruption();
+  wait();
+}
+
+void RipCDDialog::MediaChangedPollingThread::run() {
+  while (!isInterruptionRequested()) {
+    if (dialog_.ripper_->MediaChanged()) {
+      qLog(Debug) << "media changed!";
+      dialog_.ResetDialog();
+      if (dialog_.CheckCDIOIsValid()) {
+        dialog_.loader_->LoadSongs();
+      }
+    }
+    msleep(500);
+  }
+}
```

Comment:
Could this be done with a QTimer in RipCDDialog instead of a separate thread? You shouldn't need the mutex in that case.

Reply:
Just to confirm whether I understand Qt correctly: There would no longer be need for the mutex because the timer event would be handled by the main event loop, i.e., no concurrency actually happens?

Reply:
That's correct. You can connect the timer's timeout() signal to a slot and, by default, the slot code will run on the recipient's thread. It looks like there were some new callOnTimeout methods added to QTimer in 5.12, but we need to be compatible with Qt 5.7 until the Debian Stretch EOL.